### ISSUE TITLE
Network Error: ldap: response channel closed

### ISSUE THREAD

**Issue by kumy** [pinepain/ldap-auth-proxy#10]
I'm trying to protect an application wich load many .js files.

**Actual behavior**
On first app access, I'm requested to authenticate, then I also need to authenticate for each subsequent file load.

**Expected behavior**
I should authenticate once, and my authentication reused for next files access.

I can see in FF console that the first request ends with http 200, and all next requests by a 401 while there is effectively the right header in all subsequent requests `Authorization:"Basic xxxx"`.

I can see in the responses headers `X-Ldapauth-Trace:"Failed to authenticate: LDAP Result Code 200 "Network Error": ldap: response channel closed"`. They also appear a lot in `ldap-auth-proxy` logs.

(How can I send you log traces privately?)

**Comment by kumy** [comment]
![screenshot from 2018-11-07 11-55-41](https://user-images.githubusercontent.com/176794/48127567-55edac00-e284-11e8-9308-3f5bce60c33f.png)

**Comment by pinepain** [comment]
Let me try to reproduce this. A side note - my main use case is auth backend, and you are using it as proxy, right?

**Comment by kumy** [comment]
@pinepain exact

**Comment by kumy** [comment]
I've just tested using the "auth backend" pattern and it seems to work fine, but it add some configuration overhead.

Also, chain start to be long in my case: traefik -> nginx (=> ldap-auth-backend) -> app(gunicorn). It would be nice if I can suppress the nginx stack.

**Comment by kumy** [comment]
Problem is probably hidden in the "auth backend" pattern because of `proxy_cache_key "$http_authorization";` instruction.

**Comment by pinepain** [comment]
Regards nginx use with traefik, Have you looked into https://docs.traefik.io/configuration/entrypoints/#forward-authentication instead of proxying? I'm not sure traefik can cache auth response but even though it could be more effective to use traefik for proxying as ldap-auth-proxy has very primitive proxy, it's more for fast drop-in to get some auth asap, I'm not sure it's fully qualified for production use in proxy mode, but maybe it is good enough, dunno, need to look into how it's done in traefik and other, will appreciate any feedback here.

**Comment by kumy** [comment]
@pinepain didn't knew about ability for Traefik to forward authentication, I have to look this, thanks for the input!

**Comment by cosi1** [comment]
Hello, @pinepain and @kumy 

Apologies for piggybacking on this thread :-)
@kumy, have you had any success running ldap-auth-proxy with Traefik using forward authentication? I'm trying to set up LDAP authentication for my app, all bundled in a Docker Swarm stack:

```
version: "3"

services:
  xxxxx:
    image: xxxxx
    ports:
      - "3838:3838"
    user: root
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"
    deploy:
      mode: replicated
      replicas: 5
      placement:
        constraints: [node.role == worker]
      labels:
        - "traefik.port=3838"
        - "traefik.frontend.rule=PathPrefix:/"
        - "traefik.backend.loadbalancer.stickiness=true"
        - "traefik.docker.network=xxxxx_default"
        - "traefik.frontend.auth.forward.address=http://{{ ansible_host }}:8888"
        - "traefik.frontend.auth.forward.trustForwardHeader=true"
    networks:
      - default

  traefik:
    image: traefik
    command: --docker --docker.swarmMode --docker.watch --api --loglevel=DEBUG
    ports:
      - "80:80"
      - "3939:8080"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    deploy:
      mode: replicated
      replicas: 1
      placement:
        constraints: [node.role == manager]
    networks:
      - default

  ldap:
    image: pinepain/ldap-auth-proxy
    ports:
      - "8888:8888"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    deploy:
      mode: replicated
      replicas: 1
      placement:
        constraints: [node.role == manager]
    networks:
      - default
    environment:
      - LOG_LEVEL=debug
      - LISTEN=:8888
      - LDAP_SERVER=xxxxx
      - LDAP_BASE=xxxxx
      - LDAP_BIND_DN=xxxxx
      - LDAP_BIND_PASSWORD=xxxxx
      - LDAP_USER_FILTER=xxxxx
      - HEADERS_MAP=xxxxx
      - UPSTREAM=http://{{ ansible_host }}/

networks:
  default:
      driver: overlay
      driver_opts:
        com.docker.network.driver.mtu: 1450
```

However, when I authenticate with my credentials, it gets stuck, with logs flooded with the following messages:

```
time="2018-12-19T13:57:48Z" level=debug msg="Malformed auth header value"
time="2018-12-19T13:57:48Z" level=debug msg="GET / 401"
time="2018-12-19T13:57:48Z" level=debug msg="Headers: map[xxxxx], Attributes: map[xxxxx]"
```

I'm not sure if that's Traefik's fault (I suspect it removes the authorization header) or something wrong with my configuration.

**Comment by kumy** [comment]
Hi @cosi1, sorry no input from my side, as I didn't have time yet to migrate to such config, but I'm interested in a working configuration example ;)

**Comment by cosi1** [comment]
@kumy Thanks anyway! When (if) I finally make it work, I'll post my solution here :-)

@pinepain Could you please elaborate how exactly the "auth backend" mode works? If I understand correctly, Traefik expects forward authentication mechanisms to simply return a response with status code 200, without redirection. I'm not sure about the `Authorization` header, but I can't help wondering why it's *ldap-auth-proxy*, and not *Traefik* that complains about this missing header (`Malformed auth header value`). Moreover, I get `GET /` requests from ldap-auth-proxy - why??? ;-)

**Comment by pinepain** [comment]
Unfortunately, I'm not using the exact setup which suffers from this issue and I'm badly tied up with other things so I really don't know when and whether I will have a chance to look into this, so I would really appreciate PR on this issue. Sorry folks.

**Comment by cosi1** [comment]
Hopefully someone from Traefik gets interested in LDAP authentication and heads to your repository :-) I myself don't know Go, so I can't help...

**Comment by pinepain** [comment]
Hi there, I just hit this issue in k8s with ingress-nginx. During auth I create and close ldap connection on each request, which may cause troubles to the upstream LDAP server, which IMO is causing the error we seen. In a typical web app there is main page load where normally first auth occurs and then when page content loaded and parsed, all resources (css, js, images) start loading, which is a good bunch of requests within a short time. Keeping connection hanging forever was causing pita for me before (using in-house quickly written script), so this is why i'm opening it every time.

To address this issue, I'm adding caching of successfully authed creds for 5min (it's hardcoded atm, not sure it worth to change it to some other value, though, PR is welcomed). It should be addressed in #12 and available in 0.2.0 release. I will really appreciate any feedback.

**Comment by pinepain** [comment]
Closing as it should be addressed by 0.2.0 and there was no feedback that the issue occurs after that.
